internal/jobcontainers/job.go

```diff
@@ -0,0 +1,308 @@
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectExtendedLimitInformation, uintptr(unsafe.Pointer(&eliInfo)), uint32(unsafe.Sizeof(eliInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set extended limit info on job object: %s", err)
+		}
+	}
+
+	if limits.cpuRate != 0 {
+		cpuInfo := winapi.JOBOBJECT_CPU_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_CPU_RATE_CONTROL_ENABLE | winapi.JOB_OBJECT_CPU_RATE_CONTROL_HARD_CAP,
+			Rate:         limits.cpuRate,
+		}
+		_, err := windows.SetInformationJobObject(job.jobHandle, windows.JobObjectCpuRateControlInformation, uintptr(unsafe.Pointer(&cpuInfo)), uint32(unsafe.Sizeof(cpuInfo)))
+		if err != nil {
+			return fmt.Errorf("failed to set cpu limit info on job object: %s", err)
+		}
+	}
+
+	if limits.maxBandwidth != 0 || limits.maxIops != 0 {
+		ioInfo := winapi.JOBOBJECT_IO_RATE_CONTROL_INFORMATION{
+			ControlFlags: winapi.JOB_OBJECT_IO_RATE_CONTROL_ENABLE,
+		}
+		if limits.maxBandwidth != 0 {
+			ioInfo.MaxBandwidth = limits.maxBandwidth
+		}
+		if limits.maxIops != 0 {
+			ioInfo.MaxIops = limits.maxIops
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// createJobObject creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func createJobObject(name string) (*jobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
+	if err != nil {
+		return nil, err
+	}
+	iocpHandle, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 1)
+	if err != nil {
+		windows.Close(jobHandle)
+		return nil, err
+	}
+	if _, err = attachIOCP(jobHandle, iocpHandle); err != nil {
+		windows.Close(jobHandle)
+		windows.Close(iocpHandle)
+		return nil, err
+	}
+	return &jobObject{jobHandle, iocpHandle}, nil
+}
+
+// close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *jobObject) close() error {
+	closeErr := false
+	if job.jobHandle != 0 {
+		if err := windows.Close(job.jobHandle); err != nil {
+			closeErr = true
+		}
+		job.jobHandle = 0
+	}
+	if job.iocpHandle != 0 {
+		if err := windows.Close(job.iocpHandle); err != nil {
+			closeErr = true
+		}
+		job.iocpHandle = 0
+	}
+	if closeErr {
+		return errors.New("failed to close one or more handles")
+	}
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *jobObject) assign(p *JobProcess) error {
+	if p.Pid() == 0 {
+		return errors.New("process has not started")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, uint32(p.Pid()))
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
```

Comment:
Just looking at the docs, does JOB_OBJECT_ASSIGN_PROCESS get set on the handle somewhere?

Reply:
We don't pass in any security descriptors when creating the job so it has all of the access rights

Reply:
"The handle returned by CreateJobObject has JOB_OBJECT_ALL_ACCESS access to the job object."